### PR TITLE
Feature/refactoring

### PR DESCRIPTION
**Pull request by IvanSheyenko**

### CODE REVIEW

File: src/SIM.Sitecore9Installer/Events/EventManager.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SIM.Sitecore9Installer.Events
+{
+  public class EventManager
+  {
+    private static EventManager _instance=new EventManager();
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/Events/EventManager.cs:11]: missed space _instance=new

File: src/SIM.Sitecore9Installer/BaseParameters.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  public abstract class BaseParameters:IEnumerable<InstallParam>
+  {
+    private object _lock = new object();
+    private bool _evaluated;
+    public BaseParameters()
+    {
+      EventManager.Instance.ParamValueUpdated += ParamValueUpdated;
+    }
+
+    protected virtual void ParamValueUpdated(object sender, ParamValueUpdatedArgs e)
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/BaseParameters.cs:21]: I suggest you renaming method name to OnParamValueUpdated or ParamValueUpdated_EventHandler

File: src/SIM.Sitecore9Installer/BaseParameters.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      {
+        return this.Parameters.FirstOrDefault(p => p.Name == name);
+      }
+    }
+
+    public InstallParam this[int index]
+    {
+      get
+      {
+        return this.Parameters[index];
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/BaseParameters.cs:79]: possible out of range exception

File: src/SIM.Sitecore9Installer/BaseParameters.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    private object _lock = new object();
+    private bool _evaluated;
+    public BaseParameters()
+    {
+      EventManager.Instance.ParamValueUpdated += ParamValueUpdated;
+    }
+
+    protected virtual void ParamValueUpdated(object sender, ParamValueUpdatedArgs e)
+    {
+      if (this.Parameters == null)
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/BaseParameters.cs:23]: I don't see where the property is initialized. Is it supposed to be in the child class?

File: src/SIM.Sitecore9Installer/GlobalParameters.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    protected internal GlobalParameters()
+    {
+      this._parameters = new List<InstallParam>();
+    }
+
+    protected override List<InstallParam> Parameters { get => _parameters; }
+
+    protected override void CalculateParameters()
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/GlobalParameters.cs:42]: Why we need such a complicated wrapper? Why don't invoke EvaluateGlobalParams directly?

File: src/SIM.Sitecore9Installer/GlobalParameters.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        this.EvaluateGlobalParams();
+    }
+
+    public string GetGlobalParamsScript()
+    {
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/GlobalParameters.cs:48]: Why we need such a complicated wrapper? Why don't invoke GetParamsScript directly?

File: src/SIM.Sitecore9Installer/InstallParam.cs [hunk 7]
```diff
@@ -118,15 +109,29 @@ private string GetStringValue(string value)
 
       return value;
     }
-  }
 
-  public class ParamValueUpdatedArgs : EventArgs
-  {
-    public ParamValueUpdatedArgs(string oldValue)
+    private InstallParamType ParseInstallParamType(string type)
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Sitecore9Installer/InstallParam.cs:113]: I think the type parameter should be an enum type

File: src/SIM.Tool/GlobalParamsConfig/XP/GlobalParams_9.0.0_XP0.json [hunk 8]
```diff
@@ -21,32 +21,32 @@
     //xConnect
     "XConnectCert": "$GlobalParams.SqlDbPrefix+\"xconnect_client\"",
     "SqlMarketingAutomationUser": "",
-    "SqlMarketingAutomationPassword": "Password12345",
+    "SqlMarketingAutomationPassword": "Password1234567890",
```
Comment (ViacheslavAfanasyev) [inline src/SIM.Tool/GlobalParamsConfig/XP/GlobalParams_9.0.0_XP0.json:24]: I suggest adding more complex password with special symbols

File: src/SIM.Sitecore9Installer/BaseParameters.cs (hunk 4 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/BaseParameters.cs:23]: yes. The property is abstract

File: src/SIM.Sitecore9Installer/Events/EventManager.cs (hunk 1 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/Events/EventManager.cs]: fixed :)

File: src/SIM.Sitecore9Installer/BaseParameters.cs (hunk 2 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/BaseParameters.cs]: renamed to OnParamValueUpdated

File: src/SIM.Sitecore9Installer/BaseParameters.cs (hunk 3 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/BaseParameters.cs:79]: Discussed. Will not fix.

File: src/SIM.Sitecore9Installer/InstallParam.cs (hunk 7 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/InstallParam.cs:113]: No, this method parses string to enum which represents a type.

File: src/SIM.Sitecore9Installer/GlobalParameters.cs (hunk 6 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/GlobalParameters.cs:48]: This method is a useful wrapper to get correctly encoded parameter script. Will not change.

File: src/SIM.Tool/GlobalParamsConfig/XP/GlobalParams_9.0.0_XP0.json (hunk 8 above)
Comment (IvanSheyenko) [reply src/SIM.Tool/GlobalParamsConfig/XP/GlobalParams_9.0.0_XP0.json]: fixed

File: src/SIM.Sitecore9Installer/GlobalParameters.cs (hunk 5 above)
Comment (IvanSheyenko) [reply src/SIM.Sitecore9Installer/GlobalParameters.cs:42]: Fixed.

**Review by AndreyFilchenkov — APPROVED** [review]

**Pull request merged**